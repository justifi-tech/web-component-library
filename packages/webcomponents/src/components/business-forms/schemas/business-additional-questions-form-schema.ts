import { object, string } from 'yup';

export const additionQuestionsSchema = object({
  business_revenue: string().required('Enter business revenue'),
  business_payment_volume: string().required('Enter business payment volume'),
  business_dispute_volume: string().required('Enter business dispute volume'),
  business_receivable_volume: string().required('Enter business receivable volume'),
});
