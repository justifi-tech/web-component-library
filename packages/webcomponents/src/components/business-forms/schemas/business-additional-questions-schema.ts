import { object, string } from 'yup';
import { paymentVolumeValidation, revenueValidation } from './schema-validations';

export const additionalQuestionsSchema = (allowOptionalFields?: boolean) => {
  const schema = object({
    business_revenue: revenueValidation.required('Enter business revenue'),
    business_payment_volume: paymentVolumeValidation.required('Enter business payment volume'),
    business_when_service_received: string().required('Select when service is received'),
    business_recurring_payments: string().required('Select recurring payments'),
    business_recurring_payments_percentage: string().required('Enter recurring payments percentage'),
    business_seasonal: string().required('Select business seasonal'),
    business_other_payment_details: string().nullable(),
  });

  const easySchema = object({
    business_revenue: revenueValidation.nullable(),
    business_payment_volume: paymentVolumeValidation.nullable(),
    business_when_service_received: string().nullable(),
    business_recurring_payments: string().nullable(),
    business_recurring_payments_percentage: string().nullable(),
    business_seasonal: string().nullable(),
    business_other_payment_details: string().nullable(),
  });

  return allowOptionalFields ? easySchema : schema;
};
