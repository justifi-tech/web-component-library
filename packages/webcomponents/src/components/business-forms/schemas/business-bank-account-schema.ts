import { object } from 'yup';
import { 
  accountNumberValidation, 
  accountTypeValidation, 
  bankNameValidation, 
  identityNameValidation, 
  nicknameValidation, 
  routingNumberValidation } 
from './schema-validations';

export const businessBankAccountSchema = (allowOptionalFields?: boolean) => {
  const schema = object({
    bank_name: bankNameValidation.required('Enter bank name'),
    nickname: nicknameValidation.required('Enter nickname'),
    account_owner_name: identityNameValidation.required('Enter account owner name'),
    account_type: accountTypeValidation.required('Select account type'),
    account_number: accountNumberValidation.required('Enter account number'),
    routing_number: routingNumberValidation.required('Enter routing number')
  });

  const easySchema = object({
    bank_name: bankNameValidation.nullable(),
    nickname: nicknameValidation.nullable(),
    account_owner_name: identityNameValidation.nullable(),
    account_type: accountTypeValidation.nullable(),
    account_number: accountNumberValidation.nullable(),
    routing_number: routingNumberValidation.required('Enter routing number')
  });

  return allowOptionalFields ? easySchema : schema;
}
