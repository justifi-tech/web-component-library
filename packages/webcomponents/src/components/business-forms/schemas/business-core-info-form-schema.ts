import { object, string } from 'yup';

export const coreInfoSchema = object({
  legal_name: string().required('Enter legal name'),
  website_url: string()
    .url('Enter valid website url')
    .required('Enter website url'),
  email: string().email('Enter valid email').required('Enter email'),
  phone: string().required('Enter phone number'),
  doing_business_as: string().required('Enter doing business as'),
  business_type: string().required('Select business type'),
  business_structure: string().required('Select business structure'),
  industry: string().required('Enter a business industry'),
});

type BusinessType =
  | 'for_profit'
  | 'non_profit'
  | 'government_entity'
  | 'individual'
  | '';
type BusinessStructure =
  | 'sole_proprietorship'
  | 'single_llc'
  | 'multi_llc'
  | 'private_partnership'
  | 'private_corporation'
  | 'unincorporated_association'
  | 'public_partnership'
  | 'public_corporation'
  | 'incorporated'
  | 'unincorporated'
  | 'government_unit'
  | 'government_instrumentality'
  | 'tax_exempt_government_instrumentality'
  | '';

export const BusinessTypeOptions: { label: string; value: BusinessType }[] = [
  {
    label: 'Choose business type',
    value: '',
  },
  {
    label: 'Individual',
    value: 'individual',
  },
  {
    label: 'For Profit',
    value: 'for_profit',
  },
  {
    label: 'Non Profit',
    value: 'non_profit',
  },
  {
    label: 'Government Entity',
    value: 'government_entity',
  },
];

export const BusinessStructureOptions: {
  label: string;
  value: BusinessStructure;
}[] = [
    {
      label: 'Choose business structure',
      value: '',
    },
    {
      label: 'Sole Proprietorship',
      value: 'sole_proprietorship',
    },
    {
      label: 'LLC (Single)',
      value: 'single_llc',
    },
    {
      label: 'LLC (Multiple)',
      value: 'multi_llc',
    },
    {
      label: 'Private Partnership',
      value: 'private_partnership',
    },
    {
      label: 'Private Corporation',
      value: 'private_corporation',
    },
    {
      label: 'Unincorporated Association',
      value: 'unincorporated_association',
    },
    {
      label: 'Public Partnership',
      value: 'public_partnership',
    },
    {
      label: 'Public Corporation',
      value: 'public_corporation',
    },
    {
      label: 'Incorporated',
      value: 'incorporated',
    },
    {
      label: 'Unincorporated',
      value: 'unincorporated',
    },
    {
      label: 'Government Unit',
      value: 'government_unit',
    },
    {
      label: 'Government Instrumentality',
      value: 'government_instrumentality',
    },
    {
      label: 'Tax Exempt Government Instrumentality',
      value: 'tax_exempt_government_instrumentality',
    },
  ];
  