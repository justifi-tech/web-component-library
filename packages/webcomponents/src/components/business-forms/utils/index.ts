export * from './business-form-options';
export * from './helpers';
export * from './input-handlers';
export * from './event-types';
export * from './country-form-helpers';
