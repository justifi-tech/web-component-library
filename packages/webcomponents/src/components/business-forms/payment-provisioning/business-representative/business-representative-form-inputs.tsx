import { Component, h, Prop } from '@stencil/core';
import { FormController } from '../../../form/form';
import { updateAddressFormValues, updateDateOfBirthFormValues, updateFormValues } from '../../utils/input-handlers';
import { PHONE_MASKS, SSN_MASK } from '../../../../utils/form-input-masks';

@Component({
  tag: 'justifi-business-representative-form-inputs',
})
export class RepresentativeFormInputs {
  @Prop() representativeDefaultValue: any;
  @Prop() errors: any;
  @Prop() formController: FormController;

  inputHandler = (name: string, value: string) => {
    updateFormValues(this.formController, { [name]: value });
  }

  onAddressFormUpdate = (values: any): void => {
    updateAddressFormValues(this.formController, {
      ...this.formController.values.getValue().address,
      ...values,
    });
  }
  
  render() {
    return (
      <form>
        <fieldset>
          <legend>Representative</legend>
          <hr />
          <div class='row gy-3'>
            <div class='col-12 col-md-8'>
              <form-control-text
                name='name'
                label='Full Name'
                defaultValue={this.representativeDefaultValue?.name}
                errorText={this.errors.name}
                inputHandler={this.inputHandler}
              />
            </div>
            <div class='col-12 col-md-4'>
              <form-control-text
                name='title'
                label='Title'
                defaultValue={this.representativeDefaultValue?.title}
                errorText={this.errors.title}
                inputHandler={this.inputHandler}
              />
            </div>
            <div class='col-12 col-md-6'>
              <form-control-text
                name='email'
                label='Email Address'
                defaultValue={this.representativeDefaultValue?.email}
                errorText={this.errors.email}
                inputHandler={this.inputHandler}
              />
            </div>
            <div class='col-12 col-md-6'>
              <form-control-number-masked
                name='phone'
                label='Phone Number'
                defaultValue={this.representativeDefaultValue?.phone}
                errorText={this.errors.phone}
                inputHandler={this.inputHandler}
                mask={PHONE_MASKS.US}
              />
            </div>
            <div class='col-12 col-md-4'>
              <form-control-date
                name='dob_full'
                label='Birth Date'
                defaultValue={this.representativeDefaultValue?.dob_full}
                errorText={this.errors.dob_full}
                inputHandler={this.inputHandler}
                onFormControlInput={(e) => updateDateOfBirthFormValues(e, this.formController)}
              />
            </div>
            <div class='col-12 col-md-8'>
              <form-control-number-masked
                name='identification_number'
                label='SSN'
                defaultValue={this.representativeDefaultValue?.identification_number}
                errorText={this.errors.identification_number}
                inputHandler={this.inputHandler}
                mask={SSN_MASK}
              />
            </div>
            <div class='col-12'>
              <justifi-identity-address-form
                errors={this.errors.address}
                defaultValues={this.representativeDefaultValue?.address}
                handleFormUpdate={this.onAddressFormUpdate}
              />
            </div>
          </div>
        </fieldset>
      </form>
    );
  }
};
