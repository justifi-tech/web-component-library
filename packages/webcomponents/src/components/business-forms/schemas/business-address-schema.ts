import { object, string } from 'yup';
import { 
  cityValidation,
  lineOneValidation, 
  lineTwoValidation,
  postalValidation,
  stateValidation,
  makePostalValidation,
  makeStateValidation
} from './schema-validations';
import { CountryCode } from '../../../utils/country-codes';

// Base address schema (no country-specific constraints on state/postal)
export const baseAddressSchema = (allowOptionalFields?: boolean) => {
  const schema = object({
    line1: lineOneValidation.required('Enter street address'),
    line2: lineTwoValidation.nullable(),
    city: cityValidation.required('Enter city'),
    state: string().transform((v) => (v === '' ? undefined : v)).required('Select state'),
    postal_code: string().transform((v) => (v === '' ? undefined : v)).required('Enter postal code'),
    country: string().required('Select country')
  });

  const easySchema = object({
    line1: lineOneValidation.nullable(),
    line2: lineTwoValidation.nullable(),
    city: cityValidation.nullable(),
    state: string().transform((v) => (v === '' ? undefined : v)).nullable(),
    postal_code: string().transform((v) => (v === '' ? undefined : v)).nullable(),
    country: string().required('Select country')
  });

  return allowOptionalFields ? easySchema : schema;
};

// Country-specific schema convenience wrappers
export const addressSchemaUSA = (allowOptionalFields?: boolean) =>
  baseAddressSchema(allowOptionalFields).concat(object({
    state: stateValidation.required('Select state'),
    postal_code: postalValidation.required('Enter postal code'),
  } as any));

export const addressSchemaCAN = (allowOptionalFields?: boolean) =>
  baseAddressSchema(allowOptionalFields).concat(object({
    state: makeStateValidation(CountryCode.CAN).required('Select state'),
    postal_code: makePostalValidation(CountryCode.CAN).required('Enter postal code'),
  } as any));

// Back-compat default: USA
export const addressSchema = (allowOptionalFields?: boolean) => addressSchemaUSA(allowOptionalFields);
