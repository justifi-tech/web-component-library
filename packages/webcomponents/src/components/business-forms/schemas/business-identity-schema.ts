import { object, string } from 'yup';
import { addressSchema } from './business-address-schema';
import { phoneRegex } from '../utils/helpers';

export const identitySchema = (title: string, allowOptionalFields?: boolean) => {
  const schema = object({
    name: string().required(`Enter ${title} name`),
    email: string()
      .email(`Enter valid ${title} email`)
      .required(`Enter ${title} email`),
    phone: string().matches(phoneRegex, 'Enter valid phone number').required('Enter phone number'),
    dob_day: string().required(`Enter ${title} birth day`),
    dob_month: string().required(`Enter ${title} birth month`),
    dob_year: string().required(`Enter ${title} birth year`),
    identification_number: string(),
    address: addressSchema(allowOptionalFields),
  });

  const easySchema = object({
    name: string().required(`Enter ${title} name`),
    email: string()
      .email(`Enter valid ${title} email`)
      .nullable(),
    phone: string().matches(phoneRegex, 'Enter valid phone number').nullable(),
    dob_day: string().nullable(),
    dob_month: string().nullable(),
    dob_year: string().nullable(),
    identification_number: string().nullable(),
    address: addressSchema(allowOptionalFields),
  });

  return allowOptionalFields ? easySchema : schema;
}
