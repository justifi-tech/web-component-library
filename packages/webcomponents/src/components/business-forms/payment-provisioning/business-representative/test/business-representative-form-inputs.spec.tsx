import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { RepresentativeFormInputs } from '../business-representative-form-inputs';
import { FormController } from '../../../../../ui-components/form/form';
import { identitySchemaUSA, identitySchemaCAN } from '../../../schemas/business-identity-schema';
import { CountryCode } from '../../../../../utils/country-codes';

describe('justifi-business-representative-form-inputs', () => {
  const mockFormController = () => {
    return new FormController(identitySchemaUSA('representative'));
  };

  describe('Country-specific identification number label', () => {
    test('renders USA SSN label and help text', async () => {
      const formController = mockFormController();
      const representativeData = { identification_number: '', ssn_last4: null };
      
      const page = await newSpecPage({
        components: [RepresentativeFormInputs],
        template: () => (
          <justifi-business-representative-form-inputs
            formController={formController}
            representativeDefaultValue={representativeData}
            errors={{}}
            country={CountryCode.USA}
          />
        ),
      });
      
      expect(page.root).toMatchSnapshot();
      
      // Verify the identification number field uses toggleable-field with correct USA-specific label and help text
      const idField = page.root.querySelector('toggleable-field[fieldname="identification_number"]');
      expect(idField).toBeTruthy();
      expect(idField.getAttribute('label')).toBe('SSN');
      expect(idField.getAttribute('helptext')).toBe('Enter your full Social Security Number. It is required for Federal OFAC check.');
    });

    test('renders CAN SIN label and help text', async () => {
      const formController = new FormController(identitySchemaCAN('representative'));
      const representativeData = { identification_number: '', ssn_last4: null };
      
      const page = await newSpecPage({
        components: [RepresentativeFormInputs],
        template: () => (
          <justifi-business-representative-form-inputs
            formController={formController}
            representativeDefaultValue={representativeData}
            errors={{}}
            country={CountryCode.CAN}
          />
        ),
      });
      
      expect(page.root).toMatchSnapshot();
      
      // Verify the identification number field uses toggleable-field with correct Canada-specific label and help text
      const idField = page.root.querySelector('toggleable-field[fieldname="identification_number"]');
      expect(idField).toBeTruthy();
      expect(idField.getAttribute('label')).toBe('SIN');
      expect(idField.getAttribute('helptext')).toBe('Enter your full Social Insurance Number.');
    });
  });

  describe('Identification number toggle behavior', () => {
    test('shows regular input when no ssn_last4 is present', async () => {
      const formController = mockFormController();
      const representativeData = { 
        identification_number: '',
        ssn_last4: null 
      };
      
      const page = await newSpecPage({
        components: [RepresentativeFormInputs],
        template: () => (
          <justifi-business-representative-form-inputs
            formController={formController}
            representativeDefaultValue={representativeData}
            errors={{}}
            country={CountryCode.USA}
          />
        ),
      });
      
      const idField = page.root.querySelector('toggleable-field[fieldname="identification_number"]');
      expect(idField).toBeTruthy();
      expect(idField.getAttribute('readonlyvalue')).toBe(null);
      expect(idField.getAttribute('mask')).toBe('000-00-0000');
    });

    test('shows read-only display when ssn_last4 is present', async () => {
      const formController = mockFormController();
      const representativeData = { 
        identification_number: '123456789',
        ssn_last4: '6789' 
      };
      
      const page = await newSpecPage({
        components: [RepresentativeFormInputs],
        template: () => (
          <justifi-business-representative-form-inputs
            formController={formController}
            representativeDefaultValue={representativeData}
            errors={{}}
            country={CountryCode.USA}
          />
        ),
      });
      
      const idField = page.root.querySelector('toggleable-field[fieldname="identification_number"]');
      expect(idField).toBeTruthy();
      expect(idField.getAttribute('readonlyvalue')).toBe('6789');
    });
  });

  describe('Form component behavior', () => {
    test('uses toggleable-field instead of form-control-number-masked', async () => {
      const formController = mockFormController();
      const representativeData = {};
      
      const page = await newSpecPage({
        components: [RepresentativeFormInputs],
        template: () => (
          <justifi-business-representative-form-inputs
            formController={formController}
            representativeDefaultValue={representativeData}
            errors={{}}
            country={CountryCode.USA}
          />
        ),
      });
      
      const idField = page.root.querySelector('toggleable-field[fieldname="identification_number"]');
      expect(idField).toBeTruthy();
      
      // Should not find the old masked input
      const maskedIdField = page.root.querySelector('form-control-number-masked[name="identification_number"]');
      expect(maskedIdField).toBeNull();
    });
  });
});