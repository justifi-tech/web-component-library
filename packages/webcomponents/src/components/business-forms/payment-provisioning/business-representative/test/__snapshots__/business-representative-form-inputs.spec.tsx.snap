// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`justifi-business-representative-form-inputs Country-specific identification number label renders CAN SIN label and help text 1`] = `
<justifi-business-representative-form-inputs>
  <form>
    <fieldset>
      <div class="align-items-center d-flex gap-2">
        <legend class="mb-0" part="heading-2 heading text color font-family">
          Representative
        </legend>
        <form-control-tooltip helptext="The representative for your business needs to be someone who has significant control over managing your business’s finance. You will have an opportunity to add owners later in the form."></form-control-tooltip>
      </div>
      <hr class="mt-2">
      <div class="gy-3 row">
        <div class="col-12 col-md-8">
          <form-control-text label="Full Name" name="name"></form-control-text>
        </div>
        <div class="col-12 col-md-4">
          <form-control-text helptext="Role at your business, e.g. President, CEO, Treasurer." label="Title" name="title"></form-control-text>
        </div>
        <div class="col-12 col-md-6">
          <form-control-text label="Email Address" name="email"></form-control-text>
        </div>
        <div class="col-12 col-md-6">
          <form-control-number-masked label="Phone Number" mask="(000) 000-0000" name="phone"></form-control-number-masked>
        </div>
        <div class="col-12 col-md-4">
          <form-control-date helptext="Must be 18 or older." label="Birth Date" name="dob_full"></form-control-date>
        </div>
        <div class="col-12 col-md-8">
          <form-control-number-masked defaultvalue="" helptext="Enter your full Social Insurance Number." label="SIN" mask="000-000-000" name="identification_number"></form-control-number-masked>
        </div>
        <div class="col-12">
          <justifi-identity-address-form country="CAN"></justifi-identity-address-form>
        </div>
      </div>
    </fieldset>
  </form>
</justifi-business-representative-form-inputs>
`;

exports[`justifi-business-representative-form-inputs Country-specific identification number label renders USA SSN label and help text 1`] = `
<justifi-business-representative-form-inputs>
  <form>
    <fieldset>
      <div class="align-items-center d-flex gap-2">
        <legend class="mb-0" part="heading-2 heading text color font-family">
          Representative
        </legend>
        <form-control-tooltip helptext="The representative for your business needs to be someone who has significant control over managing your business’s finance. You will have an opportunity to add owners later in the form."></form-control-tooltip>
      </div>
      <hr class="mt-2">
      <div class="gy-3 row">
        <div class="col-12 col-md-8">
          <form-control-text label="Full Name" name="name"></form-control-text>
        </div>
        <div class="col-12 col-md-4">
          <form-control-text helptext="Role at your business, e.g. President, CEO, Treasurer." label="Title" name="title"></form-control-text>
        </div>
        <div class="col-12 col-md-6">
          <form-control-text label="Email Address" name="email"></form-control-text>
        </div>
        <div class="col-12 col-md-6">
          <form-control-number-masked label="Phone Number" mask="(000) 000-0000" name="phone"></form-control-number-masked>
        </div>
        <div class="col-12 col-md-4">
          <form-control-date helptext="Must be 18 or older." label="Birth Date" name="dob_full"></form-control-date>
        </div>
        <div class="col-12 col-md-8">
          <form-control-number-masked defaultvalue="" helptext="Enter your full Social Security Number. It is required for Federal OFAC check." label="SSN" mask="000-00-0000" name="identification_number"></form-control-number-masked>
        </div>
        <div class="col-12">
          <justifi-identity-address-form country="USA"></justifi-identity-address-form>
        </div>
      </div>
    </fieldset>
  </form>
</justifi-business-representative-form-inputs>
`;

exports[`justifi-business-representative-form-inputs Form component behavior uses toggleable-field instead of form-control-number-masked 1`] = `
<justifi-business-representative-form-inputs>
  <form>
    <fieldset>
      <div class="align-items-center d-flex gap-2">
        <legend class="mb-0" part="heading-2 heading text color font-family">
          Representative
        </legend>
        <form-control-tooltip helptext="The representative for your business needs to be someone who has significant control over managing your business’s finance. You will have an opportunity to add owners later in the form."></form-control-tooltip>
      </div>
      <hr class="mt-2">
      <div class="gy-3 row">
        <div class="col-12 col-md-8">
          <form-control-text label="Full Name" name="name"></form-control-text>
        </div>
        <div class="col-12 col-md-4">
          <form-control-text helptext="Role at your business, e.g. President, CEO, Treasurer." label="Title" name="title"></form-control-text>
        </div>
        <div class="col-12 col-md-6">
          <form-control-text label="Email Address" name="email"></form-control-text>
        </div>
        <div class="col-12 col-md-6">
          <form-control-number-masked label="Phone Number" mask="(000) 000-0000" name="phone"></form-control-number-masked>
        </div>
        <div class="col-12 col-md-4">
          <form-control-date helptext="Must be 18 or older." label="Birth Date" name="dob_full"></form-control-date>
        </div>
        <div class="col-12 col-md-8">
          <form-control-number-masked helptext="Enter your full Social Security Number. It is required for Federal OFAC check." label="SSN" mask="000-00-0000" name="identification_number"></form-control-number-masked>
        </div>
        <div class="col-12">
          <justifi-identity-address-form country="USA"></justifi-identity-address-form>
        </div>
      </div>
    </fieldset>
  </form>
</justifi-business-representative-form-inputs>
`;
